rep-2009.rst

```diff
@@ -0,0 +1,446 @@
+
+
+Terminology
+===========
+
+:Regular Publisher:
+  A publisher that publishes a single message type on a single topic.
+
+:Regular Subscriber:
+  A subscriber that subscribes to a single message type on a single topic.
+
+:Supported Message Type:
+  A message type that can be selected for use during negotiation.
+
+:Supported Message Priority Map:
+  A data structure that maps supported message types to their respective priority.
+  The exact function used to determine how to use these priorities can be overwritten by the user.
+  In general, higher numbers indicate higher priority, zero indicates no preference, and negative numbers indicate a vote against using a supported message type.
+
+:Selected Message Type:
+  A supported type that has been selected for use during negotiation.
+
+:Negotiating Publisher:
+  A publisher with preferences as to which supported message type it prefers to publish.
+  It can publish one or more selected message types on the same number of topics.
+
+:Negotiating Subscriber:
+  A subscriber with preferences as to which supported message type it prefers to subscribe to.
+  It subscribes to a single selected message type on a single topic.
+
+:Negotiating Publisher and Subscriber Pair:
+  A pair of negotiating publishers and subscribers that are related to each other in that the subscribers preferences may be based on the selected message type of the publisher.
+  An example would be if the subscriber receives data, modifies it, and then publishes it.
+  In this case, it is desirable to have the publisher and subscriber use the same selected message type.
+  However, the selected message type may not be known before the publisher negotiates it.
+
+Specification
+=============
+
+High-level Goals
+----------------
+
+The proposed change has a few goals:
+
+- Provide a way for nodes to publish different types of messages.
+- Support publishing in multiple formats at the same time.
+- Only necessary publishers and subscribers are active at any given time.
+- A publisher or subscriber can delay stating its preferences to wait for additional information.
+- Core functions are exposed to the user to overwrite, namely the ability to choose how priorities are used to pick the selected message types and the ability to delay stating preferences.
+
+
+Defining Supported Message Types
+--------------------------------
+
+For each supported message type, the following information must be known:
+
+- The topic that the supported message type should be published on.
+- The message type that will be sent (either a ROS 2 message type, such as  ``sensor_msgs::msg::Image``, or a custom data type).
+- The relative priority of the supported message type.
+
+The message type is important because multiple supported message types may be sent with the same ROS 2 message type.
+For example, a publisher may publish image data in the format of YUV420 or ARGB8888, both of which could be sent in a ``sensor_msgs::msg::Image``.
```

Comment:
I'm not sure I understand. 

This part doesn't rely on type adaptation. You could imagine passing the same ROS message type on two different channels, both of which are in YUV format, but say a different size image, and you choose depending on available resources. In this case, maybe you want to handle them differently and thus they could both register as supported types. I don't believe the user has to define any additional meta data than the supported type and have a code path for the message once it's received.

Reply:
So I guess this is some confusion on my part on what the "types" are.  As far as I understand, they can't just be ROS message types (like `sensor_msgs/msg/Image`), since publishing a YUV and an RGB message would use exactly the same ROS message type.  There has to be some kind of meta-type that defines a ROS message type *and* particular metadata that would be embedded in that message.

So maybe I'll change the comment to a question: how is the user expected to tell the negotiation system that there are two different "types", but they both happen to use the same ROS message type?

Reply:
I believe that I've explained this better now. See https://github.com/ros-infrastructure/rep/pull/325/commits/c88fc7e0ffe729a3a3699d6b8c0883e677cf78d3.